provider/provider.go

```diff
@@ -5,25 +5,45 @@ import (
 	"io/ioutil"
 	"strings"
 	"text/template"
+	"unicode"
 
 	"github.com/BurntSushi/toml"
 	"github.com/containous/traefik/autogen"
 	"github.com/containous/traefik/safe"
 	"github.com/containous/traefik/types"
-	"unicode"
 )
 
 // Provider defines methods of a provider.
 type Provider interface {
 	// Provide allows the provider to provide configurations to traefik
 	// using the given configuration channel.
-	Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool) error
+	Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints []*types.Constraint) error
 }
 
 // BaseProvider should be inherited by providers
 type BaseProvider struct {
-	Watch    bool
-	Filename string
+	Watch       bool
+	Filename    string
+	Constraints []*types.Constraint
+}
+
+// MatchConstraints must match with EVERY single contraint
+// returns first constraint that do not match or nil
+// returns errors for future use (regex)
+func (p *BaseProvider) MatchConstraints(tags []string) (bool, *types.Constraint, error) {
+	// if there is no tags and no contraints, filtering is disabled
+	if len(tags) == 0 && len(p.Constraints) == 0 {
+		return true, nil, nil
+	}
+
+	for _, constraint := range p.Constraints {
+		if ok := constraint.MatchConstraintWithAtLeastOneTag(tags); xor(ok == true, constraint.MustMatch == true) {
```

Comment:
`bool == true` is a bit redundant so this would be clearer to write as simply `ok != constraint.MustMatch`